lib/async_hooks.js

```diff
@@ -0,0 +1,486 @@
+  // here is critical.
+  if (async_hook_fields[kInit] === 0)
+    return;
+
+  // This can run after the early return check b/c running this function
+  // manually means that the embedder must have used initTriggerId().
+  if (!Number.isSafeInteger(triggerId)) {
+    if (triggerId !== undefined)
+      resource = triggerId;
+    triggerId = initTriggerId();
+  }
+
+  // I'd prefer allowing these checks to not exist, or only throw in a debug
+  // build, in order to improve performance.
+  if (!Number.isSafeInteger(asyncId) || asyncId < 0)
+    throw new RangeError('asyncId must be an unsigned integer');
+  if (typeof type !== 'string' || type.length <= 0)
+    throw new TypeError('type must be a string with length > 0');
+  if (!Number.isSafeInteger(triggerId) || triggerId < 0)
+    throw new RangeError('triggerId must be an unsigned integer');
+
+  for (var i = 0; i < active_hooks_array.length; i++) {
+    if (typeof active_hooks_array[i][init_symbol] === 'function') {
+      runInitCallback(
+        active_hooks_array[i][init_symbol], asyncId, type, triggerId, resource);
+    }
+  }
+
+  // Isn't null if hooks were added/removed while the hooks were running.
+  if (tmp_active_hooks_array !== null) {
+    restoreTmpHooks();
+  }
+}
+
+
+function emitBeforeN(asyncId, triggerId) {
+  for (var i = 0; i < active_hooks_array.length; i++) {
+    if (typeof active_hooks_array[i][before_symbol] === 'function') {
+      runCallback(active_hooks_array[i][before_symbol], asyncId);
+    }
+  }
+
+  if (tmp_active_hooks_array !== null) {
+    restoreTmpHooks();
+  }
+}
+
+
+// Usage: emitBeforeS(asyncId[, triggerId]). If triggerId is omitted then
+// asyncId will be used instead.
+function emitBeforeS(asyncId, triggerId) {
+  // CHECK(Number.isSafeInteger(asyncId) && asyncId > 0)
+  // CHECK(Number.isSafeInteger(triggerId) && triggerId > 0)
+
+  // Validate the ids.
+  if (asyncId < 0 || triggerId < 0) {
+    fatalError('before(): asyncId or triggerId is less than zero ' +
+               `(asyncId: ${asyncId}, triggerId: ${triggerId})`);
+  }
+
+  pushAsyncIds(asyncId, triggerId);
+
+  if (async_hook_fields[kBefore] === 0) {
+    return;
+  }
+
+  emitBeforeN(asyncId, triggerId);
+}
+
+
+// Called from native. The asyncId stack handling is taken care of there before
+// this is called.
+function emitAfterN(asyncId) {
+  if (async_hook_fields[kAfter] > 0) {
+    for (var i = 0; i < active_hooks_array.length; i++) {
+      if (typeof active_hooks_array[i][after_symbol] === 'function') {
+        runCallback(active_hooks_array[i][after_symbol], asyncId);
+      }
+    }
+  }
+
+  if (tmp_active_hooks_array !== null) {
+    restoreTmpHooks();
+  }
+}
+
+
+// TODO(trevnorris): Calling emitBefore/emitAfter from native can't adjust the
+// kIdStackIndex. But what happens if the user doesn't have both before and
+// after callbacks.
+function emitAfterS(asyncId) {
+  emitAfterN(asyncId);
+  popAsyncIds(asyncId);
```

Comment:
There's a call to `pushAsyncIds()` in `emitBeforeS()`. The push/pop of async ids here wraps the user's callback, so it's done in the `before`/`after` hooks.